### PR TITLE
Processor: `SolStakerStakeTokens`

### PR DESCRIPTION
**Pull request by febo**
This PR adds the `SolStakerStakeTokens` processor implementation. This instruction is used by `SolStakerState` to stake tokens on a specified validator.

### CODE REVIEW

File: program/src/processor/sol_staker_stake_tokens.rs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+    // sol staker stake
+    // - owner must be the stake program
+    // - must be initialized
+    // - must have the correct derivation (validates the config account)
+
+    require!(
+        ctx.accounts.sol_staker_stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "validator stake"
```
Comment (joncinque) [inline program/src/processor/sol_staker_stake_tokens.rs]: nit:
```suggestion
        "sol staker stake"
```

File: program/src/processor/sol_staker_stake_tokens.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    let limit = calculate_maximum_stake_for_sol_amount(sol_staker_stake.lamports_amount)?;
+
+    require!(
+        updated_staked_amount as u128 <= limit,
+        StakeError::TotalStakeAmountExceedsSolLimit,
+        "current staked amount ({}) + new amount ({}) exceeds limit ({})",
+        sol_staker_stake.delegation.amount,
+        amount,
+        limit
+    );
```
Comment (joncinque) [inline program/src/processor/sol_staker_stake_tokens.rs]: So just to make sure, the idea is that syncing between the SOL stake and PAL stake only ever happens in `Sync`, correct? If so, sounds great!

File: clients/rust/tests/setup/sol_staker_stake.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+};
+
+use super::stake::{create_stake_account, delegate_stake_account};
+
+pub struct SolStakerStakeManager {
+    // Stake account.
+    pub stake: Pubkey,
+    // Stake authority.
+    pub authority: Keypair,
+    // Validator account.
```
Comment (joncinque) [inline clients/rust/tests/setup/sol_staker_stake.rs]: nit:
```suggestion
    // SOL stake account.
```

File: clients/rust/tests/sol_staker_stake_tokens.rs [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+        &config_manager.mint,
+        &config_manager.vault,
+    )
+    .await;
+
+    // When we stake 6_500_000_000 tokens.
+    //
+    // - raw amount to be staked: 6_500_000_000
+    // - current lamports staked: 5_000_000_000
+    // - stake limit: 1.3 * 5_000_000_000 = 6_500_000_000
```
Comment (joncinque) [inline clients/rust/tests/sol_staker_stake_tokens.rs:80]: :heart: always hit the limits!

**Review by joncinque — APPROVED** [review]
Just tiny nits, looks great!

File: program/src/processor/sol_staker_stake_tokens.rs (hunk 2 above)
Comment (buffalojoec) [reply program/src/processor/sol_staker_stake_tokens.rs]: Agreed from my side, let's keep it simple and only do it in `Sync`.

File: program/src/processor/sol_staker_stake_tokens.rs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    require!(amount > 0, StakeError::InvalidAmount);
+
+    let updated_staked_amount = sol_staker_stake
+        .delegation
+        .amount
+        .checked_add(amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+    // maximum allowed stake based on the SOL amount
+    let limit = calculate_maximum_stake_for_sol_amount(sol_staker_stake.lamports_amount)?;
```
Comment (buffalojoec) [inline program/src/processor/sol_staker_stake_tokens.rs]: Just pointing out that `sol_amount` and `lamports_amount` are overlapping here, which might be confusing.

**Review by buffalojoec — APPROVED** [review]
Lgtm!!

File: program/src/processor/sol_staker_stake_tokens.rs (hunk 2 above)
Comment (febo) [reply program/src/processor/sol_staker_stake_tokens.rs]: Cool!

File: program/src/processor/sol_staker_stake_tokens.rs (hunk 5 above)
Comment (febo) [reply program/src/processor/sol_staker_stake_tokens.rs]: Renamed to `calculate_maximum_stake_for_lamports_amount` for clarity.

**Review by buffalojoec — APPROVED** [review]

**Pull request merged**